### ISSUE TITLE
Streamline distance configuration

### ISSUE THREAD

**Issue by lhofhansl** [luanti-org/luanti#10703]
## Problem

As is there are least the following config options that relate to viewing distance:
**Client:**
`viewing_range`
`client_mapblock_limit`

**Server:**
`max_block_send_distance`
`max_block_generate_distance`
`active_object_send_range_blocks`
`active_block_range`

In addition client and server settings not aligned. The client might have a far viewing range, but the server a low send distance, in which case far blocks would just not load.

## Solutions

1. Add a server config tab for these four settings. Then people can easily configure these at least for singleplayer.
2. Have the server send its configured send distance, so that the client can (optionally) adjust its viewing range. Might be good to have this server controlled anyway. Perhaps handle this like zoom, where there server has the option to fix this.
3. Calculate the client mapblock limit dynamically.

I think a bunch of my concerns would just go away if we did (1) above. That way we set all of these low, but it's easy (and obvious how) to increase these.

**Comment by hecktest** [comment]
The relationship between client_mapblock_limit and the draw distance isn't intuitive, and needs at least some solid documentation.
I think all the settings are there for a reason, but it would be useful to clearly define how they interact.
Messing with those settings isn't going to achieve much without an overhaul of how the client's visible range works. It's not a problem with the settings but rather with the whole data model.

**Comment by lhofhansl** [comment]
Hmm... Disregard the client limit for a moment... I am mainly concerned with the server settings, since they are very hard to identify if you do not know about them.

Extra UI treatment - the extra tab for the four server settings - makes them visible and we can explain them there, and normal MT users can change them and are not just surprised why you cannot increase the viewing_range.

Feel free to make more proposals on how to actually overhaul the client visible range.

**Comment by lhofhansl** [comment]
Perhaps some performance-presets: low, default, high, extreme (or something) that adjust viewing_range and the various other settings together. Default would be our current settings. Can't get easier then that.

**Comment by lhofhansl** [comment]
Something like this:
![image](https://user-images.githubusercontent.com/446130/101295500-d8822600-37d2-11eb-93a7-3468ffc0cd54.png)

It's located in the "start game" dialog and not the settings, because it only makes sense there for a local game. Could also have a slider.

**Comment by Zughy** [comment]
I agree on the presets, so that players can tweak MT without going deep into the settings, but I disagree on how it's showcased. I mean, the whole main menu is bad already, but aside that, I think it's kind of forced in there. I'd put it in the Settings windows, specifying somehow (tooltip?) it's for local games only

**Comment by lhofhansl** [comment]
Where would you put in the settings tab? It's also pretty cramped there.

I also tried a scrollbar. The problem with that is that if I want to show feedback in the dialog, I need to refresh the dialog (callback returns true) and then the scrollbar loses focus.

In any case, I kinda lost interest in this. Instead of focusing on the problem at hand there're discussions about how to make it perfect. Perhaps someone else can pick it up then.

@Zughy it's not your comment but mostly the discussion about distance defaults for decent machines, and the discussion on IRC, etc.

It's not even for me - I know how to set these all these settings correctly together. That was for MT beginners to avoid them getting disappointed.

**Comment by lhofhansl** [comment]
I think we can derive the other settings from 'max_block_send_distance` as follows:
- `max_block_generate_distance` = floor(`max_block_send_distance` - `chunksize / 2`)
- `active_block_range` = ceil(cbrt(`max_block_send_distance`) + 1)
- `active_object_send_range_blocks` = floor(`max_block_send_distance` * 2 / 3)

I'm conservative with the active blocks as the work required from it is significant and cubed with the distance.
The same is technically true for the active object range, but these are only in the view cone.

That would happen if all of these are set to 0, which would be the new default. **Only `max_block_send_distance` would typically be set.**

**Note that this would lead the current default and seems reasonable overall.**

In a different PR the server would also send its `max_block_send_distance` to the client, and if the client defaults its viewing_range to 0 it would set it to the server's (`max_block_send_distance` - 1) * 16. (the -1 so there's some extra buffer of blocks at the horizon that could be preloaded.

Again. **Only if all of these settings are set to 0.** Any client/server are free to manually configure these in any way they want.

**Comment by lhofhansl** [comment]
`client_mapblock_limit` = 4/3 * pi * `max_block_send_distance`^2 * 10 ~=  `max_block_send_distance`^2 * 40 (once the server sends this).
This is reasonable for small'ish distances and avoid very large numbers for larger values. (the client looks through these periodically to find which ones can be unloaded).

**Comment by hecktest** [comment]
Here we go again with the non-euclidean mapblock limits. Just use a damn sphere already, don't try to outsmart math, because math always wins.

**Comment by lhofhansl** [comment]
I knew you would come in and comment that. :)
And I will again say that any block in the view cone is never dropped.

The full sphere of cache does not make sense for large viewing_range and I will not implement or approve any PR that does that. Period. (But cool, of course if two other core-devs agree.)

Set the viewing_range/block-send-distance to 500/32 and you suddenly have to cache over 137k blocks. Need to store all these meshes, and worse need to look to through all these and see which ones expire. On the client that implies a sort operation. Perhaps if we change the implementation of Map::timerUpdate I could be amenable.

(Heck, I'd even be ok with r^3 * 1.5 - roughly the size of the view cone with a 72 degree FOV)

**Comment by Zughy** [comment]
I didn't want to demoralise you @lhofhansl , I should have come with a solution too, my bad!

Settings menu have really tall dropdown fields. If their height might be cut off of 1/3, a new dropdown field could be added in the central column

**Comment by lhofhansl** [comment]
Hey @Zughy all good.

The mantra I go by is that one or two people disagree with me I try to convince them; and if many people disagree I review my position.

Perhaps this is not as important as I think - simply because most folks do not seem to care about this much. That's all :)

**Comment by paramat** [comment]
> The client might have a far viewing range, but the server a low send distance, in which case far blocks would just not load.

This is not a problem and is actually necessary.
Imagine you have loaded-in a long thin strip of world into the client, due to send distance not being particularly high, the player will want to see long distance along this strip even though the player's view range will see further than the close edges of the strip.

The key point here is that the furthest section of world is not actually determined by server send distance, it is almost always much further away because players move.

> Have the server send its configured send distance, so that the client can (optionally) adjust its viewing range. Might be good to have this server controlled anyway. 

Auto-setting player view range to server send distance is impractical and actually undesirable for players. I consider it not worth doing even as an option.

> Add a server config tab for these four settings. Then people can easily configure these at least for singleplayer.

Seems good.
Should certainly be in the settings tab, there is more room now as we have removed some shaders, only requires a small space too. Otherwise we could make some room somehow.

> I think we can derive the other settings from 'max_block_send_distance` as follows:

:-1:

> max_block_generate_distance = max_block_send_distance - floor(chunksize / 2)

It almost always necessary and desirable to independently control these and have a different relationship. Reducing one by half a chunksize below the other is only one somewhat questionable approach i chose for default settings, it is not a general desirable relationship (i may have given the wrong impression about this elsewhere).
Please do not continue to pursue this relationship, i have already strongly objected to this elsewhere =)

> active_block_range = ceil(cbrt(max_block_send_distance) + 1)

Because ABMs are very intensive, and are often intensively used in games, it is almost always necessary and desirable to independently control this. It should not be determined by or altered by send distance.
Similarly for object send range.

I see that you mean this system to be optional, but this system is enabled by default.
What you propose is not practical or desirable, so i consider it not worth doing even as an option.

**Comment by hecktest** [comment]
**active_block_range** controls both ABMs and entities.
The description for the setting states explicitly: _This should be configured together with active_object_send_range_blocks_.

The real problem is that ABMs are somehow lumped together with entities in this one setting.

**Comment by lhofhansl** [comment]
I'm going to make a PR with presets. These are defined in Lua and hardcoded as part of the UI code. The drop box is then going to be placed in the settings dialog taking some space from the shader options. And then you can take it or leave. :)

There is surprisingly (to me) little interest in making MT work for casual users. All we do is pointing out what does not work and how it could be more perfect, and then many years after I raised this first we have quite literally **nothing**.

Proposed now are:
- presets, simple sets of config options that make sense together.
- allowing most options to be defaulted by setting them to 0 (like we did with emerge threads) along with a distance slider.
- (I'll add this) bring the distance options close together in the all-settings, maybe under a "Distance" heading, so they can be set together.

None of these are perfect. For example what happens to a preset when you manually set just one or two of the values?

@hecktest yeah. The logic is that anything active happens definitely in the ABM radius. What entities are sent to the client is controlled separately.
And - as a compromise - I added a change that allow entities in the view-cone of the player defined by FOV and active_object_send_range_blocks, so that entities wouldn't just vanish if you're a little bit away - which made MT look stupid.

**Comment by Zughy** [comment]
Sorry to bring this up again but, for the space: you could move clouds and opaque water into advanced settings to gain all the space you need while closing https://github.com/minetest/minetest/issues/8367
